### PR TITLE
Script mode horovod

### PR DESCRIPTION
**Pull request by uditbhatia**
*Description of changes:*
* `Docker.cpu` updated with Horovod & its dependencies
* `Docker.gpu` updated with Horovod & its dependencies
* `training.py` updated with Generic MPI integration in `sagemaker-containers`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: resources/change-hostname.sh [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+sed -ie "s/PLACEHOLDER_HOSTNAME/$1/g" /changehostname.c
+
+gcc -o /changehostname.o -c -fPIC -Wall /changehostname.c
+gcc -o /libchangehostname.so -shared -export-dynamic /changehostname.o -ldl
```
Comment (mvsusp) [inline resources/change-hostname.sh]: I believe this file should be in sagemaker containers.

File: resources/changehostname.c [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#include <stdio.h>
```
Comment (mvsusp) [inline resources/changehostname.c]: same

File: src/sagemaker_tensorflow_container/training.py [hunk 3]
```diff
@@ -138,7 +142,10 @@ def train(env):
 
         if not _is_host_master(env.hosts, env.current_host):
             _wait_until_master_is_down(env.hosts[0])
-
+    elif mpi_enabled:
+        logger.info("Running training job with MPI.")
+        framework.entry_point.run(env.module_dir, env.user_entry_point,
+                                  env.to_cmd_args(), env.to_env_vars(), mpi_enabled=mpi_enabled)
```
Comment (mvsusp) [inline src/sagemaker_tensorflow_container/training.py]: Do we need the statement below?

File: resources/change-hostname.sh (hunk 1 above)
Comment (uditbhatia) [reply resources/change-hostname.sh]: Done

File: src/sagemaker_tensorflow_container/training.py (hunk 3 above)
Comment (uditbhatia) [reply src/sagemaker_tensorflow_container/training.py]: Can you please elaborate? What exactly we dont need it ?

Comment (mvsusp) [reply src/sagemaker_tensorflow_container/training.py]: The logic in the conditional was repeated and the code could have being refactored.

File: docker/1.11.0/Dockerfile.cpu [hunk 4]
```diff
@@ -55,6 +57,7 @@ COPY $framework_support_installable .
 
 RUN pip install --no-cache-dir -U \
     keras==2.2.4 \
+    retrying \
```
Comment (mvsusp) [inline docker/1.11.0/Dockerfile.cpu]: sagemaker containers installs retrying, not the containers. You can remove this statement.

File: src/sagemaker_tensorflow_container/training.py [hunk 5]
```diff
@@ -138,7 +142,11 @@ def train(env):
 
         if not _is_host_master(env.hosts, env.current_host):
             _wait_until_master_is_down(env.hosts[0])
-
+    elif mpi_enabled:
```
Comment (mvsusp) [inline src/sagemaker_tensorflow_container/training.py]: Just send the enable flag, you dont need to make a conditional here.

File: test/unit/test_training.py [hunk 6]
```diff
@@ -154,6 +165,18 @@ def test_train_distributed_no_ps(run, distributed_training_env):
                            distributed_training_env.to_env_vars())
 
 
+@pytest.mark.skipif(sys.version_info.major != 3,
+                    reason="Skip this for python 2 because of dict key order mismatch")
+@patch('sagemaker_containers.beta.framework.entry_point.run')
```
Comment (mvsusp) [inline test/unit/test_training.py]: Please write tests for py2 and py3

File: docker/1.11.0/Dockerfile.cpu (hunk 4 above)
Comment (uditbhatia) [reply docker/1.11.0/Dockerfile.cpu]: done

File: src/sagemaker_tensorflow_container/training.py (hunk 5 above)
Comment (uditbhatia) [reply src/sagemaker_tensorflow_container/training.py]: done

File: test/unit/test_training.py (hunk 6 above)
Comment (uditbhatia) [reply test/unit/test_training.py]: done

**Pull request closed without merging**